drivers/UARTSerial.h

```diff
@@ -234,6 +234,18 @@ class UARTSerial : private SerialBase, public FileHandle, private NonCopyable<UA
      */
     void set_format(int bits = 8, Parity parity = UARTSerial::None, int stop_bits = 1);
 
+    /** get the available characters in the RX ringbuffer*/
+    const uint32_t get_rx_buffer_size()
```

Comment:
Despite `CircularBuffer` calling this "size", I don't find it obvious that "buffer size" means number of characters in the buffer, rather than the size of the buffer.

However, I believe in POSIX it is conventional that the "file size" indication on a device does indicate number of characters available in the RX buffer. I'll need to dig up a reference on that. Given that, maybe this should be making `FileHandle::size()` return the number, rather than adding a custom method.

The `const` on the return value is meaningless. Presumably the method itself should be `const`.

Reply:
I wanted this mainly for debugging purposes, otherwise, you have no way of checking if the ringbuffer is overflowing and overwriting characters in the beginning. with this method I was able to verify it is not overflowing, about the other things you are right!

Reply:
so should I make it uint32_t get_rx_buffer_used() const ?
or will there be a FileHandle::size() (or is there already? didnt quite understand)

Reply:
The proposal above was to use `buffer_used`. @kjbracey-arm Did you check what is the conventional name for this?

Reply:
>  I wanted this mainly for debugging purposes, otherwise, you have no way of checking if the ringbuffer is overflowing and overwriting characters in the beginning

Overwriting of the buffer never happens. `UARTSerial` checks whether the buffer is full before writing to it. So when you `write`, the return value tells you whether it was full or not. There's no possibility of writing loss. You can tell if the write buffer's full if you can't write any bytes!

For reading, yes, if the RX buffer gets full, then characters from outside are dropped, and there's no "overrun" API indication to tell you that's happened.

I guess monitoring the RX buffer level and seeing if it gets too high is a possible debugging aid, although it wouldn't be fully reliable - it would be more useful to have a proper real "overrun" API (both for hardware overrun and software buffer overrun).

And anyway, in many cases, just monitoring the returned size from non-blocking "read" would give you a hint as to the buffer level, if you're attempting big reads - if you're getting data in big chunks rather than little ones, that means there was that much data backed up in the buffer.

Seeing as we only need the RX direction, and it's only really for debugging purposes, I think I'd like it to just be "hidden" in the existing unused `size` API. It's a fairly common system extension for special devices to indicate the amount of available data in their "file size".

So just implement `UARTSerial::size()` to override `FileHandle::size()`. (At present the default `FileHandle::size` tries to call `UARTSerial::seek` to determine size, which just ends up returning an error).

This may end up being a ROM saving for everyone - that default `FileHandle::size()` is more complex than the buffer size read.